# 01. Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [2]:
# Import products.csv datafile
df_prods = pd.read_csv(r'C:\Users\David\Desktop\Instacart Basket Analysis\02 Data\Original Data\products.csv', index_col = False)

In [3]:
path = r'C:\Users\David\Desktop\Instacart Basket Analysis'

In [4]:
# Import orders.csv datafile
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03. Data consistency (df_prods)

In [5]:
# Checking for missing values in df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [6]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [7]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [8]:
df_prods.shape

(49693, 5)

In [9]:
# Addressing missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [10]:
df_prods_clean.shape

(49677, 5)

In [11]:
# Checking for duplicates in df_prods_clean
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [12]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [13]:
df_prods_clean.shape

(49677, 5)

In [14]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [15]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [16]:
# Export clean 'products' data
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# 03. Data consistency (df_ords)

In [17]:
# Perform statistical analysis for df_ords
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week    order_hour  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

Question 2 'order_hour' and 'orders_day_of_the_week' have minimum values of 0, which is strange becasue none of the variables should have a 0 value. 'order_hour' also has a max value of 2.3, which is strange.

In [18]:
# Check for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [19]:
# Check for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour                     0
days_since_prior_order    206209
dtype: int64

In [20]:
df_nan= df_ords[df_ords['days_since_prior_order'].isnull() == True]

Question 5
There are 206,209 missing values in the 'days_since_prior_order' column

In [21]:
df_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour  days_since_prior_order  
0                         2           8                     NaN  
11                        2          11                     NaN  
26                        1          14                     NaN  
39                        6          11                     NaN  
45                        3          12                     NaN  
...                     ...         ...                     ...  
3420930                   4          12                     NaN  
3420934                   3          18                     NaN  
3421002                   6          19                     NaN  
3421019                   1          15                     NaN  
3421069                   3          11                     NaN  

[206209 rows x 8 columns]

In [22]:
df_ords.shape

(3421083, 8)

In [23]:
# Addressing missing values
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   order_hour  days_since_prior_order  
0           8                     NaN  
1           7                    15.0  
2          12                    21.0  
3           7                    29.0  
4          15                    28.0

In [24]:
# Addressing missing values
df_ords['days_since_prior_order'].fillna(df_ords['days_since_prior_order'].median() , inplace=True)

In [25]:
df_ords.shape

(3421083, 8)

In [26]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   order_hour  days_since_prior_order  
0           8                     7.0  
1           7                    15.0  
2          12                    21.0  
3           7                    29.0  
4          15                    28.0

Question 6
I used the mean to replace all the missing values with the mean value becasue it would not make sense to remove aover 200,000 rows of data.

In [27]:
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [28]:
df_ords_nan

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour, days_since_prior_order]
Index: []

In [29]:
# Checking for duplicates in df_ords

df_dups = df_ords[df_ords.duplicated()]

In [30]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour, days_since_prior_order]
Index: []

Question 7
No duplicates found. 

In [31]:
df_ords_clean = df_ords

In [32]:
df_ords_clean.shape

(3421083, 8)

In [33]:
# Export clean 'orders' data
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))